### ISSUE TITLE
query (q) in subscription errors on lists

### ISSUE THREAD

**Issue by pdmangel2** [Fiware/context.Orion-LD#1248]
Following example returns 
`{
  "type": "https://uri.etsi.org/ngsi-ld/errors/BadRequestData",
  "title": "Invalid value for Subscription::q",
  "detail": "forbidden characters in String Filter"
}`

```
###
POST http://localhost:1026/ngsi-ld/v1/subscriptions
Fiware-Service: t_02
Content-Type: application/json

{
    "id": "urn:subscr:123457",
    "type": "Subscription",
    "description": "manual test",
    "entities": [
        {
            "idPattern": ".*",
            "type": "ParkingSpot"
        }
    ],
    "q": "status==\"occupied\",\"free\"",
    "watchedAttributes": ["status"],
    "notification": {
        "endpoint": {
            "uri": "http://nginx:8080/api/tools/mock/httpNoAuth",
            "accept": "application/ld+json"
        },
        "format": "keyValues",
        "attributes": [
            "status"
        ]
    }
}
```
When trying with     "q": "status==occupied,free" orion-ld crashes

**Comment by kzangeli** [comment]
ok, that's not good.
Weird, thought I had this covered by functional tests
I'll have a look at this asap!

**Comment by kzangeli** [comment]
ok, think I got it. 
There are two problems in your request:

1. "StringLists" in RHS (right-hand-side) haven't been implemented. Sorry about that, I'll get right on it.
2. The backslash isn't an allowed character (at least not in Orion-LD - I need to check this against the NGSI-LD spec).
           Meaning, you can't escape stuff in the payload body.

Try URL-encoding the q instead (and use just a single string instead of a comma-separated list):
```
"q": "status==%22occupied%22"
```

Actually, I found another problem.
I was going to propose for you to use:
```
"q": "status==%22occupied%22|status==%ssfree%22"
```
but, while the subscription is created correctly, I get no notifications fired ...

So, two bugs to fix here

**Comment by pdmangel2** [comment]
Thanks Ken!
But is '\"' not the way to escape double-quotes in JSON values ? I would expect the JSON parser then to convert escaped characters to their correct value.

**Comment by kzangeli** [comment]
Yeah, I implemented for url-encoding to have it the exact same way as URL-params.
Then we discussed it in ETSI ISG CIM some moths agio and decided we should support both url-encoding and "JSON-escaping".
I haven't implemented that last part yet.

Now, I looked at the problem a little more.
It's quite complex as Orion-LD (NGSI-LD)  attempts to be backwards-compatible with Orion (NGSIv2) and especially the "q" filter differs quite a lot.
For example, 
* string constants in NGSI-LD q have double quotes, while in NGSIv2 they don't.
* OR isn't supported in NGSIv2
* etc.

In the example functest (see PR #1249 ), I started my broker with the -experimental CLI parameter and there it works just fine. 
Both the comma-separated string list and the "same thing" but with OR instead:
```
  "q": "status==%22occupied%22,%22free%22"
  "q": "status==%22occupied2%22|status==%22free2%22"
```
But, if not in `-experimental` mode, the String list isn't implemented and ORs will never be.
As Orion-LD is moving towards the "experimental" mode (the idea is to have it more tested and then make it default mode),
fixing your problem (implementing q String-List for non-experimental mode) doesn't come very high in the backlog ...

Would you consider to move to the "experimental" mode?

Let me explain what it's all about.

Orion-LD inherits its initial code from Orion, as Orion was back in 2018 when the fork was made.
Orion then (2018) used the Mongo C++ Legacy Driver, deprecated since 2016 or so.
There's a desperate need to replace this deprecated driver with the a new supported, mongo driver,
but this is a huge effort. The entire database layer will have to be completely rewritten (mongoBackend is the library inside the broker), and as that library does FAR TOO much and the legacy driver is intertwined everywhere, the effort is even bigger.
Orion moved to "mongoc" (that's the name of the new mongo driver) about a year ago, perhaps a little more.
For Orion-LD, the work is done in a different way (way better), and I'm almost ready.
The last part is Registrations and I'm working on that right now, top priority.
Anything that is not Registrations now is implemented for mongoc (and used if the `-experimental` mode is turned on).

About 1/3 of the source code has been replaced. Lots and lots of new code, not very tested.
That's why I saved the old stuff (mongoBackend+mongo legacy driver) and put the new stuff as optional, under the `-experimental` CLI.

However, every PR comes with a functest to make sure all works and never stops working and I'm quite confident the experimental mode works pretty well. So, it's not like it's completely untested code. It's not.
There will be some bugs, of course, that's inevitable.

But, as soon as I get an issue on Orion-LD's github, I leave what I'm on and switch to bug-fixing.
Of course, if like in your case, the fix is for the old deprecated code, then I might not be willing to give it priority.
If it's instead a bug concerning the new implementation, then I drop everything else and go 100% on fixing the bug asap.

**Comment by pdmangel2** [comment]
Thanks for this elaborated explanation.
Since urgent, I have implemented a workaround : instead of specifying a 'q==...', I add the attributes I have to filter on in 'attrs' and I filter when I get updates. Much more complicated, but it works.
I certainly agree with your priorities and I'm willing to do some tests with the 'experimental' mode as soon as I get some time.
